### PR TITLE
Data Ingestion feature preview

### PR DESCRIPTION
**Pull request by itsanudeep**
Data ingestion wizard preview feature.

Note: This feature is not complete and is in preview, the code needs refactoring.

### CODE REVIEW

File: cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+const styles = (theme): StyleRules => {
+  return {
+    root: { width: '100%' },
+    pluginsRow: { display: 'flex', flexDirection: 'row', flexWrap: 'wrap' },
+    title: { display: 'flex', alignItems: 'center' },
+    pluginCard: {
+      display: 'flex',
+      flexDirection: 'column',
+      margin: 10,
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx]: modify all the pixels to use `..px`

File: cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+      justifyContent: 'center',
+    },
+  };
+};
+interface IPlugin {
+  name: string;
+  artifact: { version: string };
+  widgetJson: any;
+}
+interface ICodeEditorProps extends WithStyles<typeof styles> {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx]: why is this called CodeEditor?

File: cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+const PluginListView: React.FC<ICodeEditorProps> = ({
+  classes,
+  plugins,
+  sinks,
+  onPluginSelect,
+}) => {
+  const [anchorEl, setAnchorEl] = React.useState(null);
+
+  const open = Boolean(anchorEl);
+  const id = open ? 'simple-popover' : null;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx]: why is this needed?

File: cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      </Popover>
+    </div>
+  );
+};
+
+const StyledPluginList = withStyles(styles)(PluginListView);
+
+function PluginList(props) {
+  return (
+    <ThemeWrapper>
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginList/index.tsx]: is this being used in Angular component? if not, then this is not needed

File: cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    headingTitle: {
+      display: 'flex',
+      flexDirection: 'column',
+    },
+    pluginIcon: { width: 64 },
+  };
+};
+
+interface IPluginWidgetRendererProps extends WithStyles<typeof styles> {
+  cgProps: any;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx]: Better to expand this `configurationGroupProps`

File: cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+          <div className={classes.pluginIconContainer}>
+            <If condition={iconData}>
+              <img className={classes.pluginIcon} src={iconData} />
+            </If>
+            <If condition={!iconData}>
+              <div
+                className={`${classes.pluginFAIcon} fa ${getIcon(
+                  this.props.plugin.name.toLowerCase()
+                )}`}
+              ></div>
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx]: `</div>`

did the linter not fix this automatically?

File: cdap-ui/app/cdap/components/Ingestion/PluginsTableView/index.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                return (
+                  <span key={`${idx}-${sinkName}`}>
+                    <span
+                      className={classes.targetName}
+                      onClick={() => onSourceSinkSelect(plugin, sink)}
+                    >
+                      {sinkName}
+                    </span>
+                    <If condition={idx !== sinks.length - 1}>
+                      <span> | </span>
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginsTableView/index.tsx:152]: React will strip off extra spaces. If you need margin, you have to style it with css

File: cdap-ui/app/cdap/components/Ingestion/PluginsTableView/index.tsx [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                  <TableCell>
+                    <If condition={iconData}>
+                      <img className={classes.tablePluginIcon} src={iconData} />
+                    </If>
+                    <If condition={!iconData}>
+                      <div
+                        className={`${classes.tablePluginFAIcon} fa ${getIcon(
+                          plugin.name.toLowerCase()
+                        )}`}
+                      ></div>
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginsTableView/index.tsx]: `/>`

File: cdap-ui/app/cdap/components/Ingestion/SinkList/index.tsx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    },
+    targetsButton: {},
+  };
+};
+interface IPlugin {
+  name: string;
+  artifact: { version: string };
+  widgetJson: any;
+}
+interface ICodeEditorProps extends WithStyles<typeof styles> {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/SinkList/index.tsx]: props name

File: cdap-ui/app/cdap/components/Ingestion/SinkList/index.tsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  plugins: IPlugin[];
+  onPluginSelect: (plugin: IPlugin) => void;
+}
+
+const SinkListView: React.FC<ICodeEditorProps> = ({ classes, plugins, onPluginSelect }) => {
+  function cardSelect(plugin) {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/SinkList/index.tsx]: any other things planned to be done inside this function? otherwise just call `onPluginSelect` directly

File: cdap-ui/app/cdap/components/Ingestion/index.tsx [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+    deployFailed: false,
+    cdapVersion: '',
+    currentNs: 'default',
+  };
+
+  public componentDidMount() {
+    const currentNs = getCurrentNamespace();
+    let cdapVersion = VersionStore.getState().version;
+    if (!cdapVersion) {
+      MyCDAPVersionApi.get().subscribe((res) => {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/index.tsx:138]: is this necessary?

File: cdap-ui/app/cdap/components/Ingestion/index.tsx [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+        scope: plugin.artifact.scope,
+        keys: `widgets.${plugin.name}-${plugin.type}`,
+      };
+      const widgetJ = await MyPipelineApi.fetchWidgetJson(params).toPromise();
+      const widgetJson = JSON.parse(widgetJ[`widgets.${plugin.name}-${plugin.type}`]);
+      const pluginWithWidgetJson = { ...plugin, widgetJson };
+      return pluginWithWidgetJson;
+    });
+
+    Promise.all(promises).then((data) => {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/index.tsx:181]: all the APIs are already observable. You can use combineLatest instead of converting everything to promises

**Review by elfenheart — COMMENTED** [review]
Would be good to extract common function like getting display name. Also try to separate render and utitlity functions

File: cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx (hunk 6 above)
Comment (itsanudeep) [reply cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx]: It did not. fixed it.

File: cdap-ui/app/cdap/components/Ingestion/index.tsx (hunk 12 above)
Comment (itsanudeep) [reply cdap-ui/app/cdap/components/Ingestion/index.tsx:181]: Added a TODO, will cleanup in the upcoming days (post 6.2.0).

File: cdap-ui/app/cdap/components/Ingestion/index.tsx (hunk 11 above)
Comment (itsanudeep) [reply cdap-ui/app/cdap/components/Ingestion/index.tsx:138]: Without this, version was empty sometimes. Eg. when visiting the feature via URL/Refreshing.

File: cdap-ui/app/cdap/api/pipeline.js [hunk 13]
```diff
@@ -30,6 +30,11 @@ const extensionsPath =
   '/namespaces/:namespace/artifacts/:parentArtifact/versions/:version/extensions/:extension';
 const pluginsPath = `${extensionsPath}/plugins/:pluginName`;
 
+const extensionsFetchBase =
+  '/namespaces/:namespace/artifacts/:pipelineType/versions/:version/extensions';
+const pluginFetchBase = extensionsFetchBase + '/:extensionType';
```
Comment (yukiej) [inline cdap-ui/app/cdap/api/pipeline.js]: nit: Template strings might improve readability here

Comment (itsanudeep) [reply cdap-ui/app/cdap/api/pipeline.js]: done

File: cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            margin="dense"
+            fullWidth={true}
+            value={configurationGroupProps.label}
+            onChange={(event) => configurationGroupProps.onLabelChange(event.target.value)}
+          />
+        </div>
+        <If condition={configurationGroupProps.pluginProperties}>
+          <ConfigurationGroup
+            pluginProperties={
+              configurationGroupProps.pluginProperties && configurationGroupProps.pluginProperties
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx]: What is this intending to do?

Comment (itsanudeep) [reply cdap-ui/app/cdap/components/Ingestion/PluginWidgetRenderer/index.tsx]: Ahhh! I messed up clearing this after I added the If component. Fixed.

**Review by elfenheart — APPROVED** [review]
👍 LGTM
